### PR TITLE
Update error crates [ECR-4314]

### PR DESCRIPTION
**Pull request by slowli**
## Overview

This PR updates error libraries used by the crates in the workspace, from `failure` to `anyhow` + `thiserror`. Since we use the external derive crate for `ProtobufConvert` which requires `failure::Error` internally, the integration of `anyhow` is *slightly* hacky.

---

See: https://jira.bf.local/browse/ECR-4314

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ exonum-supervisor = { version = "1.0.0-rc.1", path = "../services/supervisor" }
 exonum-system-api = { version = "1.0.0-rc.1", path = "../components/system-api" }
 exonum-rust-runtime = { version = "1.0.0-rc.1", path = "../runtimes/rust" }
 
-failure = "0.1"
+failure = { package = "anyhow", version = "1.0" }
```
Comment (alekseysidorov) [inline cli/Cargo.toml]: Nice hack 🥇 

But are you sure that it doesn't lead to misunderstandings in the end users code?

File: components/merkledb/src/indexes/proof_list/proof.rs [hunk 2]
```diff
@@ -14,9 +14,10 @@
 
 pub use crate::ValidationError; // TODO Change for a type alias after EJB switching to rust > 1.36
 
+use anyhow as failure; // FIXME: remove once `ProtobufConvert` derive is improved (ECR-4314)
```
Comment (aleksuss) [inline components/merkledb/src/indexes/proof_list/proof.rs]: Is it really correct issue number ? Or you've meant to improve in a scope of this PR ?

File: components/merkledb/src/indexes/proof_map/proof.rs [hunk 3]
```diff
@@ -102,15 +102,15 @@ impl<'de> serde::Deserialize<'de> for ProofPath {
 }
 
 /// An error returned when a map proof is invalid.
-#[derive(Debug, Fail)]
+#[derive(Debug, Error)]
 #[non_exhaustive]
 pub enum MapProofError {
     /// Non-terminal node for a map consisting of a single node.
-    #[fail(display = "non-terminal node as a single key in proof")]
+    #[error("non-terminal node as a single key in proof")]
```
Comment (aleksuss) [inline components/merkledb/src/indexes/proof_map/proof.rs]: I'm not sure really, but I'd add a `the` article before `proof`.

File: components/merkledb/src/migration.rs [hunk 4]
```diff
@@ -583,14 +584,14 @@ impl MigrationHelper {
 }
 
 /// Errors emitted by `MigrationHelper` methods.
-#[derive(Debug, Fail)]
+#[derive(Debug, Error)]
 pub enum MigrationError {
     /// Failed to merge migration changes to database.
-    #[fail(display = "Failed to merge migration changes to database: {}", _0)]
-    Merge(#[fail(cause)] crate::Error),
+    #[error("Failed to merge migration changes to database: {}", _0)]
```
Comment (aleksuss) [inline components/merkledb/src/migration.rs]: The same here. `the` before database ?

File: components/merkledb/src/indexes/proof_list/proof.rs (hunk 2 above)
Comment (slowli) [reply components/merkledb/src/indexes/proof_list/proof.rs]: I've meant for the Jira issue to correspond to this PR and another PR dealing with this problem, but I can create a new issue just for this FIXME.

**Pull request merged**